components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc

```diff
@@ -225,9 +234,126 @@ TEST_F(PsstTabWebContentsObserverUnitTest, ShouldOnlyProcessHttpOrHttps) {
   }
 }
 
-// TODO(bridiver) - test multiple main frame navigations in a row
-// TODO(bridiver) - test redirects
-// TODO(bridiver) - test for navigating after OnLoad but before script
+TEST_F(PsstTabWebContentsObserverUnitTest,
+       ShouldProcessMultipleMainFrameNavigations) {
+  const std::string first_nav_user_script = "user1";
+  const std::string second_nav_user_script = "user2";
+  const std::string policy_script = "policy";
+  const GURL first_navigation_url("https://example1.com");
+  const GURL second_navigation_url("https://example2.com");
+
+  base::RunLoop first_nav_check_loop;
+  EXPECT_CALL(psst_rule_registry(), CheckIfMatch(first_navigation_url, _))
+      .WillOnce(CheckIfMatchCallback(
+          &first_nav_check_loop,
+          CreateMatchedRule(first_nav_user_script, policy_script)));
+
+  base::RunLoop second_nav_check_loop;
+  EXPECT_CALL(psst_rule_registry(), CheckIfMatch(second_navigation_url, _))
+      .WillOnce(CheckIfMatchCallback(
+          &second_nav_check_loop,
+          CreateMatchedRule(second_nav_user_script, policy_script)));
+
+  base::RunLoop first_nav_user_script_insert_loop;
+  EXPECT_CALL(scripts_handler(), InsertScriptInPage(first_nav_user_script, _))
+      .WillOnce(InsertScriptInPageCallback(&first_nav_user_script_insert_loop,
+                                           base::Value()));
+  base::RunLoop second_nav_user_script_insert_loop;
+  EXPECT_CALL(scripts_handler(), InsertScriptInPage(second_nav_user_script, _))
+      .WillOnce(InsertScriptInPageCallback(&second_nav_user_script_insert_loop,
+                                           base::Value()));
+  EXPECT_CALL(scripts_handler(), InsertScriptInPage(policy_script, _)).Times(0);
+
+  content::NavigationSimulator::NavigateAndCommitFromBrowser(
+      web_contents(), first_navigation_url);
+  DocumentOnLoadObserver observer(web_contents());
+  content::NavigationSimulator::NavigateAndCommitFromBrowser(
+      web_contents(), second_navigation_url);
+  observer.Wait();
+  second_nav_check_loop.Run();
+  first_nav_check_loop.Run();
+  first_nav_user_script_insert_loop.Run();
+  second_nav_user_script_insert_loop.Run();
+}
+
+TEST_F(PsstTabWebContentsObserverUnitTest, ShouldProcessRedirectsNavigations) {
+  const std::string user_script = "user";
+  const std::string policy_script = "policy";
+  const GURL url("https://example1.com");
```

Comment:
this should be a different url that also has an entry in the registry to ensure that it only picks the scripts for the redirect target